updater/internal/file/file.go

```diff
@@ -43,21 +43,42 @@ func CopyFile(logger *zap.Logger, pathIn, pathOut string, overwrite bool) error
 
 	pathOutClean := filepath.Clean(pathOut)
 	fileMode := fs.FileMode(0600)
-
 	flags := os.O_CREATE | os.O_WRONLY
 	if overwrite {
 		// If we are OK to overwrite, we will truncate the file on open
 		flags |= os.O_TRUNC
 
-		// Save old file's permissions and delete it first if it exists
-		fileOutInfo, _ := os.Stat(pathOutClean)
-		if fileOutInfo != nil {
-			fileMode = fileOutInfo.Mode()
+		// Try to save old file's permissions
+		outFileInfo, _ := os.Stat(pathOutClean)
+		if outFileInfo != nil {
+			fileMode = outFileInfo.Mode()
+		} else if useInFilePermBackup {
+			// Use the new file's permissions as a backup and don't fail on error (best chance for rollback)
+			inFileInfo, err := inFile.Stat()
+			if err != nil {
+				logger.Error("failed to retrieve fileinfo for input file", zap.Error(err))
+			}
+			if inFileInfo != nil {
+				fileMode = inFileInfo.Mode()
+			}
+		}
+
+		// Remove old file to prevent issues with mac
+		if err = os.Remove(pathOutClean); err != nil {
+			logger.Debug("Failed to remove output file", zap.Error(err))
 		}
-		_ = os.Remove(pathOutClean)
 	} else {
 		// This flag will make OpenFile error if the file already exists
 		flags |= os.O_EXCL
+
+		// Use the new file's permissions and fail if there's an issue (want to fail for backup)
+		inFileInfo, err := inFile.Stat()
+		if err != nil {
+			return fmt.Errorf("failed to retrive fileinfo for input file: %w", err)
+		}
+		if inFileInfo != nil {
+			fileMode = inFileInfo.Mode()
+		}
```

Comment:
Since the err is returning you can just change this to

```go
inFileInfo, err := inFile.Stat()
if err != nil {
    return fmt.Errorf("failed to retrive fileinfo for input file: %w", err)
}
fileMode = inFileInfo.Mode()
		
```

Reply:
same comment as above. i'll change it.